ballerina/variable_validator.bal

```diff
@@ -154,4 +172,47 @@ class VariableValidator {
             self.errors.push(getErrorDetailRecord(message, location)); 
         }
     }
+
+    isolated function checkVariableUsageCompatibility(__InputValue[] input, parser:VariableDefinition varDef,
+                                                      parser:ArgumentNode argNode) {
+        __InputValue? inputValue = getInputValueFromArray(input, argNode.getName());
+        __Type? actualType = self.getTypeRecordFromTypeName(varDef.kind);
+        if inputValue is __InputValue && actualType is __Type {
+            if getOfType(inputValue.'type)?.name != getOfType(actualType)?.name {
+                string message = string`Variable "${<string>argNode.getVariableName()}" of type "${varDef.kind}" `+
+                string`used in position expecting type "${getTypeNameFromType(inputValue.'type)}".`;
+                self.errors.push(getErrorDetailRecord(message, argNode.getLocation()));
+            } else if inputValue.'type.kind == NON_NULL && actualType.kind != NON_NULL && varDef?.defaultValue is () {
+                string message = string`Variable "${<string>argNode.getVariableName()}" of type "${varDef.kind}" `+
+                string`used in position expecting type "${getTypeNameFromType(inputValue.'type)}".`;
+                self.errors.push(getErrorDetailRecord(message, argNode.getLocation()));
```

Comment:
Using `getOfType()` here will create bugs. We should check the exact type against each other. **_IIRC, we discussed this offline and since the types should be structurally equal, we agreed to use structural equivalency directly in this case_**. Stripping the types using `getOfType` and checking the names instead of the type value, will waste that.

These complex checks are redundant in that case too.